File: 16-3sum-closest/3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        """
        one way we can do this is sort the array so that we can use ptrs to manipulate
        the closest target
        """

        nums.sort()
        cand = 0
        closest = float('inf')
        for i in range(len(nums)):
            l,r = i+1,len(nums)-1

            while l < r:
                three_sum = nums[i] + nums[l] + nums[r]

                if abs(three_sum - target) < closest:
                    closest = abs(three_sum - target)
                    cand = three_sum
                
                if three_sum > target:
                    r -= 1
                else:
                    l += 1
        
        return cand

File: 16-3sum-closest/3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int closest = Integer.MAX_VALUE;

        for(int i = 0; i < nums.length -2; i++){

            // skip dupes
            if(i > 0 && nums[i] == nums[i-1]){
                continue;
            }

            int left = i + 1;
            int right = nums.length - 1;

            while(left < right){
                int threeSum = nums[i] + nums[left] + nums[right];
                //if the difference is less than the previous one set it as the new one
                if(Math.abs(threeSum - target) < Math.abs(closest - target)){
                    closest = threeSum;
                }

                //shift pointers based on magnitude
                if(threeSum > target){
                    right--;
                }else if(threeSum < target){
                    left++;
                }else{ //return exact
                    return threeSum;
                }
            }
        }

        return closest;
    }
}